src/finn/transformation/fpgadataflow/convert_to_hls_layers.py

```diff
@@ -489,3 +491,113 @@ def apply(self, model):
+                if ll_const is not None:
+                    ll_cinit = model.get_initializer(ll_const)
+                else:
+                    continue
+
+                ll_cinit_shape = list(ll_cinit.shape)
+                # get number of channels from input
+                ll_in_layout = model.get_tensor_layout(ll_input)
+                if ll_in_layout == DataLayout.NHWC or ll_in_layout == DataLayout.NC:
+                    ch = ll_in_shape[-1]
+                elif ll_in_layout == DataLayout.NCHW:
+                    ch = ll_in_shape[1]
+                else:
+                    continue
+
+                # check if the shape of initializer is compatible with
+                # number of channels, e.g. (ch,1) or (ch)
+                # TODO: verify plausible shapes
+                if np.prod(ll_cinit_shape) != ch:
+                    continue
+
+                # check initializer contains integers as floats
+                if not (ll_cinit.astype(np.int32) == ll_cinit).all():
+                    continue
+
+                # all initializer conditions are met
+                # check inputs
+                idt = model.get_tensor_datatype(ll_input)
+
+                # skip conversion for layers with float input
+                if not idt.is_integer():
+                    continue
+
+                # check layout of inputs/outputs, and convert if needed
+                # check layout and convert if necessary
+                if ll_in_layout == DataLayout.NCHW:
+                    ll_input = nchw_to_nhwc(ll_input, model, node_ind)
+                    node_ind += 1
+                    ll_in_shape = model.get_tensor_shape(ll_input)
+
+                # keep track of where we need to insert the HLS Op
+                # it has to be ahead of the output transform
+                insert_point = node_ind
+                ll_output_layout = model.get_tensor_layout(ll_output)
+                if ll_output_layout == DataLayout.NCHW:
+                    ll_output = nchw_to_nhwc(ll_output, model, node_ind, reverse=True)
+                    node_ind += 1
+
+                # create node with no parallelization first
+                pe = 1
+                assert ch % pe == 0, "Requirement IFC divisable by PE is violated."
+
+                # set function and determine output data type
+                if node.op_type == "Add":
+                    func = "add"
+                    if idt.signed():
+                        odt = DataType.get_smallest_possible(2 * idt.min())
+                    else:
+                        odt = DataType.get_smallest_possible(2 * idt.max())
+                elif node.op_type == "Mul":
+                    func = "mul"
+                    if idt.signed():
+                        odt = DataType.get_smallest_possible(abs(idt.min()) * idt.min())
+                    else:
+                        odt = DataType.get_smallest_possible(idt.max() * idt.max())
```

Comment:
shouldn't you be also checking the data type for the parameter (ll_cinit's datatype) when deciding the output datatype? might make sense to skip any add/muls that don't have integer params too.

Reply:
Yes this code assumes the parameter has the same datatype as the input, which was the case when I initially wrote it, but not true any more